Context aware policies docs. (#169)
kubewarden/docs: docs/writing-policies/spec/host-capabilities/06-kubernetes.md

cluster where they are running on. For that, the Kubewarden SDKs expose functions
that uses the WaPC communication protocol to talk with the host system asking for
data about the cluster.

## WaPC protocol contract

Under the hood, the SDKs functions uses  two functions available in the runtime
that handle this kind of  information request. They are:

- `list_resources_by_namespace` : WaPC function to request resources information from a given namespace
- `list_resources_all`: WaPC function to request resources information from the whole cluster

This is the description of the input and output payloads used by the WaPC protocol used to expose this capability:

<table>
<tr>
<th> WaPC function name </th> <th> Input payload </th> <th> Output payload </th>
</tr>
<tr>
<td>

`list_resources_all`

</td>
<td>

```hcl
{
	# API Group version. Use `v1` for the `core` group and `groupName/groupVersion` for all other groups
	"api_version": string,
	# Resource kind
	"kind": string,
	# Label selector to filter the resources
	"label_selector": string,
	# Field selector to filter the resources
	"field_selector": string
}
```

</td>
<td>

> Result of <code>GET /apis/$api_version/$kind</code>

Reviewer: ```suggestion
Return a Kubernetes [`List`](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds), which is a collection of Kubernetes objects of the same type.
```

Reviewer: I would also add a `info` block (I hope it works here given all the funky stuff we're doing with the html table) that states that this API has to be used to get cluster-wide resources like Namespace

```hcl
{
	"apiVersion": string,
	# List type of the resource type requested (e.g. DeploymentList)
	"kind": string,
	"items":[
	# Item of the resource type requested (e.g. Deployment)
		{
			"apiVersion": string,
			"kind": string,
			"metadata":{
				"name": string,
				"namespace": string
			}
			# Field of the type requested
			# ...
		}
	],
	"metadata":{}
}
```

</td>
</tr>
<tr>
<td>

`list_resources_by_namespace`

</td>
<td>

```hcl
{
	# API Group version. Use `v1` for the `core` group and `groupName/groupVersion` for all other groups
	"api_version": string,
	# Resource kind
	"kind": string,
	# Namespace where the requested resource lives in
	"namespace": string,
	# Label selector to filter the resources
	"label_selector": string,
	# Field selector to filter the resources
	"field_selector": string
}
```

</td>
<td>

> Result of <code>GET /apis/$api_version/$kind</code>

Reviewer: ```suggestion
Return a Kubernetes [`List`](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds), which is a collection of Kubernetes objects of the same type.
```

Reviewer: I would also add a `caution` block (I hope it works here given all the funky stuff we're doing with the html table) that states that this API should NOT be used to retrieve cluster-wide resources, like Namespace.

---
sidebar_label: "Kubernetes Capabilities"
title: ""
---

# Kubernetes Capabilities

Kubewarden context aware policies requires access to resources from the Kubernetes
cluster where they are running on. For that, the Kubewarden SDKs expose functions

> that uses the waPC communication protocol to talk with the host system asking for

Reviewer: ```suggestion
that use the waPC communication protocol to talk with the host system asking for
```

data about the cluster.

## waPC protocol contract

> Under the hood, the SDKs functions rely on these functions exposed by

Reviewer: ```suggestion
Under the hood, the SDKs rely on these functions exposed by
```

the policy host environment:

> - `list_resources_by_namespace` : waPC function to request resources information from a given namespace
> - `list_resources_all`: waPC function to request resources information from the whole cluster
> - `get_resource`:  WaPC function to request a resource information by its name

Reviewer: ```suggestion
- `list_resources_by_namespace` : given a resource type and a namespace, list all the resources of that type that are defined inside of it. This cannot be used to list cluster-wide resources, like `Namespace`.
- `list_resources_all`: given a resource type, list all the resources of that type that are defined inside the whole cluster. This can be used to list cluster-wide resources, like `Namespace`.
- `get_resource`:  find the exact resource identified by a the given resource type, given name and an optional namespace identifier.
```
